test/test_sum_product.py

```diff
@@ -86,21 +87,1006 @@ def test_partition(inputs, dims, expected_num_components):
+
+    f1 = random_tensor(OrderedDict({}))
+
+    f2 = random_tensor(OrderedDict({
+        "w": Bint[w_dim],
+        "sequences": Bint[sequences],
+        "days": Bint[days],
+        "tones": Bint[tones],
+        "x_prev": Bint[x_dim],
+        "x_curr": Bint[x_dim],
+    }))
+
+    f3 = random_tensor(OrderedDict({
+        "w": Bint[w_dim],
+        "sequences": Bint[sequences],
+        "weeks": Bint[weeks],
+        "y_prev": Bint[y_dim],
+        "y_curr": Bint[y_dim],
+    }))
+
+    factors = [f1, f2, f3]
+    plate_to_step = dict({
+        "sequences": {},
+        "tones": {},
+        "days": {"x_prev": "x_curr"},
+        "weeks": {"y_prev": "y_curr"}
+    })
+
+    factors1 = modified_partial_sum_product(sum_op, prod_op, factors, vars1, plate_to_step)
+    factors2 = modified_partial_sum_product(sum_op, prod_op, factors1, vars2, plate_to_step)
+    actual = reduce(prod_op, factors2)
+
+    local_var_dict = {
+        "days": frozenset(),
+        "weeks": frozenset()
+    }
+    local_markov_var_dict = {
+        "days": frozenset(),
+        "weeks": frozenset()
+    }
+    global_vars = frozenset({"w"})
+    markov_to_step = dict({
+        "days": {"x"},
+        "weeks": {"y"},
+    })
+
+    expected = _expected_modified_partial_sum_product(
+        sum_op, prod_op, factors, plate_to_step, global_vars,
+        local_var_dict, local_markov_var_dict, markov_to_step)
+
+    assert_close(actual, expected, atol=5e-4, rtol=5e-4)
+
+
+@pytest.mark.parametrize('vars1,vars2', [
+    (frozenset(),
+     frozenset({"sequences", "time", "x_prev", "x_curr", "tones", "y_prev", "y_curr"})),
+    (frozenset({"tones", "y_prev", "y_curr"}),
+     frozenset({"sequences", "time", "x_prev", "x_curr"})),
+    (frozenset({"time", "x_prev", "x_curr", "tones", "y_prev", "y_curr"}),
+     frozenset({"sequences"})),
+    (frozenset({"sequences", "time", "tones", "x_prev", "x_curr", "y_prev", "y_curr"}),
+     frozenset()),
+])
+@pytest.mark.parametrize('x_dim,y_dim,sequences,time,tones', [
+    (2, 3, 2, 3, 2),
+    (1, 3, 2, 3, 2),
+    (2, 1, 2, 3, 2),
+    (2, 3, 2, 1, 2),
+])
+@pytest.mark.parametrize('sum_op,prod_op', [(ops.logaddexp, ops.add), (ops.add, ops.mul)])
+def test_modified_partial_sum_product_14(sum_op, prod_op, vars1, vars2,
+                                         x_dim, y_dim, sequences, time, tones):
+
+    f = random_tensor(OrderedDict({}))
+
+    g = random_tensor(OrderedDict({
+        "sequences": Bint[sequences],
+        "time": Bint[time],
+        "x_prev": Bint[x_dim],
+        "x_curr": Bint[x_dim],
+    }))
+
+    h = random_tensor(OrderedDict({
+        "sequences": Bint[sequences],
+        "time": Bint[time],
+        "tones": Bint[tones],
+        "x_curr": Bint[x_dim],
+        "y_prev": Bint[y_dim],
+        "y_curr": Bint[y_dim],
+    }))
+
+    factors = [f, g, h]
+    plate_to_step = dict({
+        "sequences": {},
+        "time": {"x_prev": "x_curr"},
+        "tones": {"y_prev": "y_curr"}
```

Comment:
I guess the reason this doesn't fail as intractable is that you haven't included `y` in the `time` plate, so it's treating `tones` as nested entirely within a single step of `time`.

Reply:
For this to fail as intractable, you might have to include two markov variables `y_prev_tones` and `y_prev_time` for `y` in `h`, one for each time variable, and put them in `plate_to_step` appropriately:
```py
plate_to_step = dict({
    "sequences": {},
    "time": {"x_prev": "x_curr", "y_time_prev": "y_curr"},
    "tones": {"y_prev": "y_curr", "y_tones_prev": "y_curr"}
})
```

Reply:
See `test_modified_partial_sum_product_15`